docs/proposals/20220206-quota.md

```diff
@@ -0,0 +1,257 @@
+
+## Motivation
+
+* Per namespace quota admission control is not suitable for multi-tenant virtual cluster environment. This is because It requires tenant to coordinate with super cluster admin to create namespaces and setup quotas in advance. It also forces super cluster admin to map namespaces with tenants in order to control overall quota usage for each tenant.
+* Cross namespace and per tenant overall quota admission control provides a true virtual cluster experience for tenant. It allows tenant freely creating/updating/deleting namespaces with guarantee of not breaking  resource quota allowance in super cluster. It also reduce resource fragmentation in tenant space. 
+* Cross namespace quota admission control needs a centralized engine to validate resource creation. Traditional approach is to use validation Webhook. However, using Webhook in multi-tenant cluster environment has following issues:
+    * for a Webhook installed in tenant virtual cluster: as tenants have admin privileges in their virtual cluster, the quota validation Webhook can be deleted by tenant and is not a truly enforced mechanism.
+    * for a Webhook installed in super cluster: it becomes a bottle neck for all tenants and impacts multiple threading model implemented in pod syncer.  In addition the Webhook needs to map namespaces with tenants, and other information originally kept in syncer to performance validation, which is not efficient.
+* Given quota admission control is a proprietary implementation for different cloud environments. Providing an universal solution is not pratical. This is why in this document we propose to build a plugin to let users to easily integrate their quota validation logics into pod-syncer.
+
+### Goals
+
+* Propose a quota validation plugin for user to integrate proprietary quota management functions into pod-syncer.
+* Define per tenant locking mechanism in pod-syncer to preserve parallel processing of tenants’ requests and prevent race condition while checking quota usage.
+* Give an example of quota validation plugin registration and initialization.
+
+### Non-Goals
+
+* Queuing mechanism used for managing pods that exceeds quota limit.
+* Define design and implementations for quota provision, quota validation and quota usage tracking.
+
+## Proposal
+
+### User Stories and Requirements
+
+* Enforce tenant overall quota usage in super cluster but allow tenant freely creating namespace, and managing separate quota in virtual cluster.
+* Enable cross namespace quota admission control in super cluster to reduce resource fragementation.
+* Allow cloud admin to define own quota validation implementation and easy integration into pod-syncer
+* Handle tenant’s resource requests in parallel and minimize interferences of large batch requests from noisy tenant.
```

Comment:
Can we write these more like:

**As a _____** _(cluster admin/operator/etc)_ **I want ________** _(feature eg 'be provide a custom quota handler that stops syncing')_ **so that _____** _(reason, eg 'I can customize what pods sync to the super cluster')_